pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,433 @@
+				controller.enqueueGateway(cur)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func (c *MCPodRouteController) createPodInformer() {
+	listOptions := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermNotEqualSelector("spec.nodeName", c.nodeConfig.Name).String()
+	}
+	c.podInformer = coreinformers.NewFilteredPodInformer(
+		c.k8sClient,
+		metav1.NamespaceAll,
+		0,
+		nil,
+		listOptions,
+	)
+	c.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				c.enqueuePod(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				c.enqueuePod(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				c.enqueuePod(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	c.podLister = corelisters.NewPodLister(c.podInformer.GetIndexer())
+}
+
+func (c *MCPodRouteController) enqueueGateway(obj interface{}) {
+	_, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueueGateway can't process event", "obj", obj)
+		return
+	}
+	c.gwQueue.Add(dummyKey)
+}
+
+// validatePod validates the Pod info to determine if it's a valid Pod which should
+// have	Openflow flows installed for cross-cluster traffic.
+func (c *MCPodRouteController) validatePod(pod *corev1.Pod) bool {
+	// pod.Status.PodIP and HostIP should be an IPv4 address for IPv4 and dual stack clusters.
+	// IPv6 only cluster is not supported by Antrea Multi-cluster yet.
+	podIP := pod.Status.PodIP
+	podNodeIP := pod.Status.HostIP
+
+	if pod.Spec.NodeName == c.nodeConfig.Name {
+		klog.V(2).InfoS("Skip Pod on the current Node")
+		return false
+	}
+
+	if podNodeIP == "" || pod.Spec.HostNetwork {
+		klog.V(2).InfoS("Pod has no valid Node IP or it is in HostNetwork, skip it", "name", klog.KObj(pod),
+			"nodeIP", podNodeIP, "hostNetwork", pod.Spec.HostNetwork)
+		return false
+	}
+
+	ips, exists := c.getPodNodeIPs(podIP)
+	if exists && ips.podNodeIP == podNodeIP && ips.podIP == podIP {
+		klog.V(2).InfoS("No Pod change impacts installed Openflow rules", "name", klog.KObj(pod))
+		return false
+	}
+	return true
+}
+
+func (c *MCPodRouteController) enqueuePod(obj interface{}, isDelete bool) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueuePod can't process event", "obj", obj)
+		return
+	}
+
+	podIP := pod.Status.PodIP
+	if podIP == "" {
+		return
+	}
+
+	podName := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+	klog.InfoS("Processing Pod", "Pod", podName.String(), "podIP", podIP)
+
+	podSet := c.getPodsByIP(podIP)
+	if podSet == nil && isDelete {
+		return
+	}
+
+	if podSet != nil && isDelete {
+		if podSet.Has(podName.String()) {
+			if podSet.Len() == 1 {
+				c.deletePodNodeIP(podIP)
+			}
```

Comment:
When there are other Pods with the same Pod IP, we may change the effective Node IP to the next Pod?

Reply:
I am not sure I got your question here, If there are other Pods with the same Pod IP(when the old one is not deleted yet),  then it will go to following codes to update the IP/NodeIP mapping. We do update the Node IP.

Reply:
I guess you have an assumption that the Pod is created/updated first must be deleted first too? E.g. if we receive: `pod1 update -> pod2 update`; then `pod1 delete` must come before `pod2 delete`, and `pod2 delete -> pod1 delete` cannot happen.

I am not very sure if that is always the case. Do you know? Why?

If `pod2 delete -> pod1 delete` can happen, then at `pod2 delete`, we should change to use pod1's Node IP.

Reply:
Are you assuming that both pod1 and pod2 are running with the same Pod IP but in different Nodes? I am not sure this case will happen. In network policy only mode, each Node has its own IP pool which won't be duplicate with other Nodes.
In other modes, I suppose each Node manages its own Pods with a Pod CIDR which should have no overlapping with other Nodes as well. I think the case is when two Pods are on the same Node maybe assigned with a duplicate IP.
@tnqn  do you have any comment for this? thanks.

Reply:
I feel the processing in `checkPod` and `enqueuePod` and the several caches are not necessary.

We could just install the Pod flow based on the eventual state when `syncPod` is called. Since we assume multiple Pods may have same IP at one moment in some cases, we could just add one podIPIndex to podIndexer.
For Pod event handlers, just enqueue the PodIP when PodIP or HostIP changes.
For `syncPod`, pseudocode is as below:
```
func (c *MCPodRouteController) syncPod(podIP string) error {
	c.podWorkersStartedMutex.RLock()
	defer c.podWorkersStartedMutex.RUnlock()
	if !c.podWorkersStarted {
		return nil
	}

	pods, _ := c.podInformer.GetIndexer().ByIndex("podIP", podIP)
	if len(pods) == 0 {
		c.ofClient.UninstallMulticlusterPodFlows(podIP)
		return nil
	}
        // Install flow for the newest Pod if there are multiple Pods having this IP.
	pod := getNewestPod(pods)
	nodeIP := pod.Status.HostIP
	c.ofClient.InstallMulticlusterPodFlows(net.ParseIP(podIP), net.ParseIP(podNodeIP))
	return nil
}
```
This should cover all cases and is simpler and less error-prone.

Reply:
How if the Pod loses its IP or the IP changes (e.g. Pod recreated)? I feel we should at least have a cache of Pod -> Pod IP.

But I also feel it is complex and not efficient to lock/unlock and read/check/write maps multiple times, and believe we can simplify without wrapping all map operations.

Reply:
If a Pod loses its IP or its IP changes, the Pod event handler still knows the old IP and the new IP, so we could enqueue both IPs. Then syncPod for the old IP could uninstall old flows and syncPod for the new IP could installs new flows.

Reply:
Enqueuing old Pod IP sounds good!

Reply:
Refine this part to use cache indexer.